### PR TITLE
hooks

### PR DESCRIPTION
**Pull request by slajax**
### overview

The generator should be able to generate standalone hooks in the current directory or by mounting them onto an existing `service.json`. When generating it in standalone mode it would be nice if it also produced the rest of the repo dependencies such as `package.json` etc so it can be published but this should not be the default implementation necessarily.

### tasks
- [x] refactor our generator pipeline utils
- [x] build out the generator.js pipeline
- [x] copy to appropriate `hooks/` dir
- [x] properly render with handlebars
- [x] rename hooks based on argv
- [x] mount to `service.json` as before / after
- [x] make sure hook runs
- [x] make sure tests run
- [x] clean up console.logs in tests (`NODE_ENV = testing`)

### dependencies
- `--config` and `--path` refactor: feathersjs/feathers-cli#50

### tech debt
- deal with linting #37 
- deal with config conflicts & validation #36

### CODE REVIEW

**Comment by slajax** [comment]
Updated issue with execution plan.

**Comment by slajax** [comment]
screen recording 2016-11-30 at 01 17 pm

**Comment by slajax** [comment]
Latest work flow can be seen here: http://io.kc.io/iJ1L

File: src/utils/mount.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        debug(err.stack);
+        return done(err);
+      }
+      debug(`Successfully mounted "${options.name}" hook to service at ${serviceConfigPath}`);
+      debug(`Service config can be found at ${relativeServiceConfigPath}`);
+      done();
+    });
+
+    done();
+  };
```
Comment (slajax) [inline src/utils/mount.js]: Can probably refactor a bunch of this to be more streamlined, but don't want to do too much of that until I've laid out the happy path for all of the different generations.

File: src/utils/rename.js [hunk 2]
```diff
@@ -33,3 +35,21 @@ export default function (options) {
...
+          let newName = file.replace('hook', options.name);
+          debug(`Renaming template ${file} to ${newName}`);
+          files[newName] = files[file];
+          delete files[file];
+        }
+      }
+    );
+
+    done();
+  };
```
Comment (slajax) [inline src/utils/rename.js]: Same here, the duplication can be minimized, but will wait until I've landed happy path on all generations and then refactor with the full picture in view.

File: src/app/templates/server/middleware/logger.js [hunk 3]
```diff
@@ -1,18 +1,22 @@
...
-        winston.info(error.stack);
+        if (process.env.NODE_ENV !== 'testing') {
+          winston.info(message);
+        }
+      } else {
+        if (process.env.NODE_ENV !== 'testing') {
+          winston.error(message);
+          winston.info(error.stack);
+        }
       }
```
Comment (slajax) [inline src/app/templates/server/middleware/logger.js]: Will refactor this with an improved default solution for logging. Winston is great, but I feel like we can do better then this and even make it configurable for adapters like logstash.

Comment (ekryski) [inline src/app/templates/server/middleware/logger.js]: Yeah. Even this is a good start though and the winston docs are pretty decent, but agreed I think we can improve this slightly. We've also been noticing that we need a better profiling/logging solution for sockets.

Comment (slajax) [inline src/app/templates/server/middleware/logger.js]: I've got a few ideas for how we can improve this to support real time logging once we have an Elastic Search adapter. Winston is fine for now, we mostly just need to support more consistency with how it's being used vs debug.

File: src/service/templates/service.filter.js [hunk 4]
```diff
@@ -1,13 +1,17 @@
...
 
-  return function() {
-    console.log(`{{options.name}} filter returned: true`);
+  log(`You are using the default filter for the \`{{options.name}}\` service (in ${__dirname}).`);
+  log(`This means all clients will get every real-time event.`);
+  log(`For more information how to filter events see http://docs.feathersjs.com/real-time/filtering.html`);
+  log(``);
+
+  return function () {
+    log(`{{options.name}} filter returned: true`);
```
Comment (ekryski) [inline src/service/templates/service.filter.js]: We'll probably want to return false by default to keep things locked down. Right now it's easy to leak events without knowing especially when people just copy-paste code.

File: src/utils/mount.js (hunk 1 above)
Comment (ekryski) [inline src/utils/mount.js]: Ya man. Future for sure but splitting out into a util function is a good first step though.

**Comment by ekryski** [comment]
Oh man. Looks great! :shipit:

File: src/service/templates/service.filter.js (hunk 4 above)
Comment (ekryski) [inline src/service/templates/service.filter.js]: @slajax ya it's a tradeoff either way. For sure people have apps in the wild that are leaking info right now so I'd almost prefer to go more secure first in this instance. Thoughts @feathersjs/core-team?

Comment (ekryski) [inline src/service/templates/service.filter.js]: In reality you should be filtering all your data anyway. Broadcasting to every connected socket is no good. I'll be happy to update the socket emitting to the new proposal so it is faster and more secure by default.

Comment (slajax) [inline src/service/templates/service.filter.js]: @ekryski can you point me at the new proposal? Happy to update the filtering to be more secure, that's definitely the best practice, I just want to make sure that when generations occur, it takes nothing more then an `npm start` to see it work. Right now the logging that occurs on `npm start` to warn you that you are using defaults is what we're using to make sure that users are aware of this default.

**Pull request merged**